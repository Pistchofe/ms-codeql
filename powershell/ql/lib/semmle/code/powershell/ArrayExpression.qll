import powershell

class ArrayExpr extends @array_expression, Expr {
  override SourceLocation getLocation() { array_expression_location(this, result) }

  StmtBlock getStmtBlock() { array_expression(this, result) }

  /**
   * Gets the i'th element of this `ArrayExpr`, if this can be determined statically.
   *
   * See `getStmtBlock` when the array elements are not known statically.
   */
  Expr getElement(int i) {
    result =
      unique( | | this.getStmtBlock().getAStmt()).(CmdExpr).getExpr().(ArrayLiteral).getElement(i)
  }

  /**
   * Gets an element of this `ArrayExpr`, if this can be determined statically.
   *
   * See `getStmtBlock` when the array elements are not known statically.
   */
  Expr getAnElement() { result = this.getElement(_) }

  override string toString() { result = "@(...)" }
}
