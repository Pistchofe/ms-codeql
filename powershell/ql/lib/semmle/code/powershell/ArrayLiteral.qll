import powershell

class ArrayLiteral extends @array_literal, Expr {
  override SourceLocation getLocation() { array_literal_location(this, result) }

  Expr getElement(int index) { array_literal_element(this, index, result) }

  Expr getAnElement() { array_literal_element(this, _, result) }

  override string toString() { result = "...,..." }
}
