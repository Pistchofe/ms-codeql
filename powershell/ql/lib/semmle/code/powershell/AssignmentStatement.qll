import powershell

class AssignmentStatement extends @assignment_statement instanceof Statement
{
    SourceLocation getLocation()
    {
        assignment_statement_location(this, result)
    }

    int getKind()
    {
        assignment_statement(this, result, _, _)
    }

    Expression getLeftHandSide()
    {
        assignment_statement(this, _, result, _)
    }

    Statement getRightHandSide()
    {
        assignment_statement(this, _, _, result)
    }

    string toString(){
        result = "AssignmentStatement at: " + this.getLocation().toString()
    }
}