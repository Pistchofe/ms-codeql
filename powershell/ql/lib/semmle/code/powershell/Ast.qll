import powershell

class Ast extends @ast {
  string toString() { none() }

  Ast getParent() { parent(this, result) } // TODO: Flip parent and child in relation in the extractor

  Location getLocation() { none() }
}
