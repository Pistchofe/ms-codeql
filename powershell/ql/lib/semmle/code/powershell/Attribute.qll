import powershell

class Attribute extends @attribute, AttributeBase {
  override string toString() { result = this.getName() }

  override SourceLocation getLocation() { attribute_location(this, result) }

  string getName() { attribute(this, result, _, _) }

  int getNumNamedArguments() { attribute(this, _, result, _) }

  int getNumPositionalArguments() { attribute(this, _, _, result) }

  NamedAttributeArgument getNamedArgument(int i) { attribute_named_argument(this, i, result) }

  NamedAttributeArgument getANamedArgument() { result = this.getNamedArgument(_) }

  int getNumberOfArguments() { result = count(this.getAPositionalArgument()) }

  Expr getPositionalArgument(int i) { attribute_positional_argument(this, i, result) }

  Expr getAPositionalArgument() { result = this.getPositionalArgument(_) }

  int getNumberOfPositionalArguments() { result = count(this.getAPositionalArgument()) }
}
