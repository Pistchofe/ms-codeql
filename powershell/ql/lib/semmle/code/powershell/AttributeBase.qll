import powershell

class AttributeBase extends @attribute_base, Ast { }
