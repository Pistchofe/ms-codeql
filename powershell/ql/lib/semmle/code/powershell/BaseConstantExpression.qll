import powershell

class BaseConstantExpression extends @base_constant_expression instanceof Expression
{
    string toString()
    {
        none()
    }
}