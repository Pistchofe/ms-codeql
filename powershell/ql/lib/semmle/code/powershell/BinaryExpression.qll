import powershell

class BinaryExpr extends @binary_expression, Expr {
  override SourceLocation getLocation() { binary_expression_location(this, result) }

  int getKind() { binary_expression(this, result, _, _) }

  /** Gets an operand of this binary expression. */
  Expr getAnOperand() {
    result = this.getLeft()
    or
    result = this.getRight()
  }

  /** Holds if this binary expression has the operands `e1` and `e2`. */
  predicate hasOperands(Expr e1, Expr e2) {
    e1 = this.getLeft() and
    e2 = this.getRight()
    or
    e1 = this.getRight() and
    e2 = this.getLeft()
  }

  Expr getLeft() { binary_expression(this, _, result, _) }

  Expr getRight() { binary_expression(this, _, _, result) }
}

abstract private class AbstractArithmeticExpr extends BinaryExpr { }

final class ArithmeticExpr = AbstractArithmeticExpr;

class AddExpr extends AbstractArithmeticExpr {
  AddExpr() { this.getKind() = 40 }

  final override string toString() { result = "...+..." }
}

class SubExpr extends AbstractArithmeticExpr {
  SubExpr() { this.getKind() = 41 }

  final override string toString() { result = "...-..." }
}

class MulExpr extends AbstractArithmeticExpr {
  MulExpr() { this.getKind() = 37 }

  final override string toString() { result = "...*..." }
}

class DivExpr extends AbstractArithmeticExpr {
  DivExpr() { this.getKind() = 38 }

  final override string toString() { result = ".../..." }
}

class RemExpr extends AbstractArithmeticExpr {
  RemExpr() { this.getKind() = 39 }

  final override string toString() { result = "...%..." }
}

abstract private class AbstractBitwiseExpr extends BinaryExpr { }

final class BitwiseExpr = AbstractBitwiseExpr;

class BitwiseAndExpr extends AbstractBitwiseExpr {
  BitwiseAndExpr() { this.getKind() = 56 }

  final override string toString() { result = "...&..." }
}

class BitwiseOrExpr extends AbstractBitwiseExpr {
  BitwiseOrExpr() { this.getKind() = 57 }

  final override string toString() { result = "...|..." }
}

class BitwiseXorExpr extends AbstractBitwiseExpr {
  BitwiseXorExpr() { this.getKind() = 58 }

  final override string toString() { result = "...^..." }
}

class ShiftLeftExpr extends AbstractBitwiseExpr {
  ShiftLeftExpr() { this.getKind() = 97 }

  final override string toString() { result = "...<<..." }
}

class ShiftRightExpr extends AbstractBitwiseExpr {
  ShiftRightExpr() { this.getKind() = 98 }

  final override string toString() { result = "...>>..." }
}

abstract private class AbstractComparisonExpr extends BinaryExpr { }

final class ComparisonExpr = AbstractComparisonExpr;

abstract private class AbstractCaseInsensitiveComparisonExpr extends AbstractComparisonExpr { }

final class CaseInsensitiveComparisonExpr = AbstractCaseInsensitiveComparisonExpr;

abstract private class AbstractCaseSensitiveComparisonExpr extends AbstractComparisonExpr { }

final class CaseSensitiveComparisonExpr = AbstractCaseSensitiveComparisonExpr;

class EqExpr extends AbstractCaseInsensitiveComparisonExpr {
  EqExpr() { this.getKind() = 60 }

  final override string toString() { result = "... -eq ..." }
}

class NeExpr extends AbstractCaseInsensitiveComparisonExpr {
  NeExpr() { this.getKind() = 61 }

  final override string toString() { result = "... -ne ..." }
}

class CEqExpr extends AbstractCaseSensitiveComparisonExpr {
  CEqExpr() { this.getKind() = 76 }

  final override string toString() { result = "... -ceq ..." }
}

class CNeExpr extends AbstractCaseSensitiveComparisonExpr {
  CNeExpr() { this.getKind() = 77 }

  final override string toString() { result = "... -cne ..." }
}

abstract private class AbstractRelationalExpr extends AbstractComparisonExpr { }

final class RelationalExpr = AbstractRelationalExpr;

abstract private class AbstractCaseInsensitiveRelationalExpr extends AbstractRelationalExpr { }

final class CaseInsensitiveRelationalExpr = AbstractCaseInsensitiveRelationalExpr;

abstract private class AbstractCaseSensitiveRelationalExpr extends AbstractRelationalExpr { }

final class CaseSensitiveRelationalExpr = AbstractCaseSensitiveRelationalExpr;

class GeExpr extends AbstractCaseInsensitiveRelationalExpr {
  GeExpr() { this.getKind() = 62 }

  final override string toString() { result = "... -ge ..." }
}

class GtExpr extends AbstractCaseInsensitiveRelationalExpr {
  GtExpr() { this.getKind() = 63 }

  final override string toString() { result = "... -gt ..." }
}

class LtExpr extends AbstractCaseInsensitiveRelationalExpr {
  LtExpr() { this.getKind() = 64 }

  final override string toString() { result = "... -lt ..." }
}

class LeExpr extends AbstractCaseInsensitiveRelationalExpr {
  LeExpr() { this.getKind() = 65 }

  final override string toString() { result = "... -le ..." }
}

class CGeExpr extends AbstractCaseSensitiveRelationalExpr {
  CGeExpr() { this.getKind() = 78 }

  final override string toString() { result = "... -cge ..." }
}

class CGtExpr extends AbstractCaseSensitiveRelationalExpr {
  CGtExpr() { this.getKind() = 79 }

  final override string toString() { result = "... -cgt ..." }
}

class CLtExpr extends AbstractCaseSensitiveRelationalExpr {
  CLtExpr() { this.getKind() = 80 }

  final override string toString() { result = "... -clt ..." }
}

class CLeExpr extends AbstractCaseSensitiveRelationalExpr {
  CLeExpr() { this.getKind() = 81 }

  final override string toString() { result = "... -cle ..." }
}

class LikeExpr extends AbstractCaseInsensitiveComparisonExpr {
  LikeExpr() { this.getKind() = 66 }

  final override string toString() { result = "... -like ..." }
}

class NotLikeExpr extends AbstractCaseInsensitiveComparisonExpr {
  NotLikeExpr() { this.getKind() = 67 }

  final override string toString() { result = "... -notlike ..." }
}

class MatchExpr extends AbstractCaseInsensitiveComparisonExpr {
  MatchExpr() { this.getKind() = 68 }

  final override string toString() { result = "... -match ..." }
}

class NotMatchExpr extends AbstractCaseInsensitiveComparisonExpr {
  NotMatchExpr() { this.getKind() = 69 }

  final override string toString() { result = "... -notmatch ..." }
}

class ReplaceExpr extends AbstractCaseInsensitiveComparisonExpr {
  ReplaceExpr() { this.getKind() = 70 }

  final override string toString() { result = "... -replace ..." }
}

abstract class AbstractTypeExpr extends BinaryExpr { }

final class TypeExpr = AbstractTypeExpr;

abstract class AbstractTypeComparisonExpr extends AbstractTypeExpr { }

final class TypeComparisonExpr = AbstractTypeComparisonExpr;

class IsExpr extends AbstractTypeComparisonExpr {
  IsExpr() { this.getKind() = 92 }

  final override string toString() { result = "... -is ..." }
}

class IsNotExpr extends AbstractTypeComparisonExpr {
  IsNotExpr() { this.getKind() = 93 }

  final override string toString() { result = "... -isnot ..." }
}

class AsExpr extends AbstractTypeExpr {
  AsExpr() { this.getKind() = 94 }

  final override string toString() { result = "... -as ..." }
}

abstract private class AbstractContainmentExpr extends BinaryExpr { }

final class ContainmentExpr = AbstractContainmentExpr;

abstract private class AbstractCaseInsensitiveContainmentExpr extends AbstractContainmentExpr { }

final class CaseInsensitiveContainmentExpr = AbstractCaseInsensitiveContainmentExpr;

class ContainsExpr extends AbstractCaseInsensitiveContainmentExpr {
  ContainsExpr() { this.getKind() = 71 }

  final override string toString() { result = "... -contains ..." }
}

class NotContainsExpr extends AbstractCaseInsensitiveContainmentExpr {
  NotContainsExpr() { this.getKind() = 72 }

  final override string toString() { result = "... -notcontains ..." }
}

class InExpr extends AbstractCaseInsensitiveContainmentExpr {
  InExpr() { this.getKind() = 73 }

  final override string toString() { result = "... -in ..." }
}

class NotInExpr extends AbstractCaseInsensitiveContainmentExpr {
  NotInExpr() { this.getKind() = 74 }

  final override string toString() { result = "... -notin ..." }
}

abstract private class AbstractLogicalBinaryExpr extends BinaryExpr { }

final class LogicalBinaryExpr = AbstractLogicalBinaryExpr;

class LogicalAndExpr extends AbstractLogicalBinaryExpr {
  LogicalAndExpr() { this.getKind() = 53 }

  final override string toString() { result = "... -and ..." }
}

class LogicalOrExpr extends AbstractLogicalBinaryExpr {
  LogicalOrExpr() { this.getKind() = 54 }

  final override string toString() { result = "... -or ..." }
}

class LogicalXorExpr extends AbstractLogicalBinaryExpr {
  LogicalXorExpr() { this.getKind() = 55 }

  final override string toString() { result = "... -xor ..." }
}

class JoinExpr extends BinaryExpr {
  JoinExpr() { this.getKind() = 59 }

  final override string toString() { result = "... -join ..." }
}

class SequenceExpr extends BinaryExpr {
  SequenceExpr() { this.getKind() = 33 }

  final override string toString() { result = "[..]" }
}

class FormatExpr extends BinaryExpr {
  FormatExpr() { this.getKind() = 50 }

  final override string toString() { result = "... -f ..." }
}
