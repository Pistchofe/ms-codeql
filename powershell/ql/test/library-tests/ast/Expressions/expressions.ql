import powershell

query predicate binaryExpr(BinaryExpr e, Expr e1, Expr e2) {
    e1 = e.getLeft() and
    e2 = e.getRight()
}

query predicate cmdExpr(CmdExpr cmd, Expr e) {
    e = cmd.getExpression()
}

query predicate invokeMemoryExpression(InvokeMemberExpression invoke, Expr e, int i, Expr arg) {
    e = invoke.getExpression() and
    arg = invoke.getArgument(i)
}

query predicate expandableString(ExpandableStringExpression expandable, int i, Expr e) {
    e = expandable.getExpr(i)
}
