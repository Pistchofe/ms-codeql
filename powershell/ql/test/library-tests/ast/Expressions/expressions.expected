binaryExpr
| BinaryExpression.ps1:3:11:3:24 | ...+... | BinaryExpression.ps1:3:11:3:16 | val1 | BinaryExpression.ps1:3:19:3:24 | val2 |
| TernaryExpression.ps1:1:9:1:16 | ...+... | TernaryExpression.ps1:1:9:1:10 | 6 | TernaryExpression.ps1:1:15:1:16 | 7 |
cmdExpr
| BinaryExpression.ps1:1:9:1:10 | 1 | BinaryExpression.ps1:1:9:1:10 | 1 |
| BinaryExpression.ps1:2:9:2:10 | 2 | BinaryExpression.ps1:2:9:2:10 | 2 |
| BinaryExpression.ps1:3:11:3:24 | ...+... | BinaryExpression.ps1:3:11:3:24 | ...+... |
| BinaryExpression.ps1:4:1:4:8 | result | BinaryExpression.ps1:4:1:4:8 | result |
| ExpandableString.ps1:1:1:1:40 | ExpandableStringExpression at: ExpandableString.ps1:1:1:1:40 | ExpandableString.ps1:1:1:1:40 | ExpandableStringExpression at: ExpandableString.ps1:1:1:1:40 |
| ExpandableString.ps1:1:23:1:38 | (no string representation) | file://:0:0:0:0 | (no string representation) |
| SubExpression.ps1:1:1:1:24 | ArrayExpression at: SubExpression.ps1:1:1:1:24 | SubExpression.ps1:1:1:1:24 | ArrayExpression at: SubExpression.ps1:1:1:1:24 |
| SubExpression.ps1:2:1:2:22 | ArrayExpression at: SubExpression.ps1:2:1:2:22 | SubExpression.ps1:2:1:2:22 | ArrayExpression at: SubExpression.ps1:2:1:2:22 |
| TernaryExpression.ps1:1:8:1:23 | ...?...:... | TernaryExpression.ps1:1:8:1:23 | ...?...:... |
| TernaryExpression.ps1:1:9:1:16 | ...+... | TernaryExpression.ps1:1:9:1:16 | ...+... |
invokeMemoryExpression
| SubExpression.ps1:1:1:1:24 | ArrayExpression at: SubExpression.ps1:1:1:1:24 | file://:0:0:0:0 | (no string representation) | 0 | SubExpression.ps1:1:21:1:23 | 10 |
expandableString
| ExpandableString.ps1:1:1:1:40 | ExpandableStringExpression at: ExpandableString.ps1:1:1:1:40 | 0 | ExpandableString.ps1:1:8:1:13 | name |
| ExpandableString.ps1:1:1:1:40 | ExpandableStringExpression at: ExpandableString.ps1:1:1:1:40 | 1 | file://:0:0:0:0 | (no string representation) |
